core/src/test/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationActionTests.java

```diff
@@ -0,0 +1,280 @@
+
+    public void testGlobalBlock() {
+        Request request = new Request();
+        PlainActionFuture<Response> listener = new PlainActionFuture<>();
+        ClusterBlocks.Builder block = ClusterBlocks.builder()
+                .addGlobalBlock(new ClusterBlock(1, "", false, true, RestStatus.SERVICE_UNAVAILABLE, ClusterBlockLevel.ALL));
+        clusterService.setState(ClusterState.builder(clusterService.state()).blocks(block));
+        try {
+            action.new AsyncSingleAction(request, listener).start();
+            listener.get();
+            fail("expected ClusterBlockException");
+        } catch (Throwable t) {
+            if (ExceptionsHelper.unwrap(t, ClusterBlockException.class) == null) {
+                logger.info("expected ClusterBlockException  but got ", t);
+                fail("expected ClusterBlockException");
+            }
+        }
+    }
+
+    public void testBasicRequestWorks() throws InterruptedException, ExecutionException, TimeoutException {
+        Request request = new Request().index("test");
+        request.shardId = 0;
+        PlainActionFuture<Response> listener = new PlainActionFuture<>();
+        clusterService.setState(ClusterStateCreationUtils.state("test", randomBoolean(), ShardRoutingState.STARTED));
+        action.new AsyncSingleAction(request, listener).start();
+        assertThat(transport.capturedRequests().length, equalTo(1));
+        transport.handleResponse(transport.capturedRequests()[0].requestId, new Response());
+        listener.get();
+    }
+
+    public void testRetryOnTimeout() throws InterruptedException, ExecutionException, TimeoutException {
+        // transport service that throws exception each time we send a request
+        transportService = new TransportService(transport, THREAD_POOL) {
+            @Override
+            public <T extends TransportResponse> void sendRequest(DiscoveryNode node, String action, TransportRequest request, TransportRequestOptions options, TransportResponseHandler<T> handler) {
+                TransportException e;
+                int excpetionType = randomInt(4);
+                switch (excpetionType) {
+                    case 1: {
+                        e = new ConnectTransportException(node, "test exception");
+                        break;
+                    }
+                    case 2: {
+                        e = new TransportException("test exception", new IndexShardClosedException(new ShardId("test", 0), "test exception"));
+                        break;
+                    }
+                    case 3: {
+                        e = new TransportException("test exception", new NodeClosedException(node));
+                        break;
+                    }
+                    case 0: {
+                        e = new TransportException("test exception", new NoShardAvailableActionException(new ShardId("test", 0)));
+                        break;
+                    }
+                    case 4: {
```

Comment:
this test is about timeouts, right? feels unnatural to throw a normal exception here. You can also rename to the test to say "handle exceptions" and check all exception are handled correctly (timeout or thrown) - imho a deidcate non-timed out error is better (i.e., it tests it returns immediately and not after a short timeout).